File: binding/src/multimedia/libqcameradevice.cpp
#include <QByteArray>
#include <QCameraDevice>
#include <QCameraFormat>
#include <QSize>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <qcameradevice.h>
#include "libqcameradevice.h"
#include "libqcameradevice.hxx"

QCameraFormat* QCameraFormat_new() {
	 return new QCameraFormat();
}

QCameraFormat* QCameraFormat_new2(const QCameraFormat* other) {
	 return new QCameraFormat(*other);
}

void QCameraFormat_OperatorAssign(QCameraFormat* self, const QCameraFormat* other) {
	self->operator=(*other);
}

int QCameraFormat_PixelFormat(const QCameraFormat* self) {
	return self->pixelFormat();
}

QSize* QCameraFormat_Resolution(const QCameraFormat* self) {
	return new QSize(self->resolution());
}

float QCameraFormat_MinFrameRate(const QCameraFormat* self) {
	return self->minFrameRate();
}

float QCameraFormat_MaxFrameRate(const QCameraFormat* self) {
	return self->maxFrameRate();
}

bool QCameraFormat_IsNull(const QCameraFormat* self) {
	return self->isNull();
}

bool QCameraFormat_OperatorEqual(const QCameraFormat* self, const QCameraFormat* other) {
	return self->operator==(*other);
}

bool QCameraFormat_OperatorNotEqual(const QCameraFormat* self, const QCameraFormat* other) {
	return self->operator!=(*other);
}

void QCameraFormat_Delete(QCameraFormat* self) {
    delete self;
}

QCameraDevice* QCameraDevice_new() {
	 return new QCameraDevice();
}

QCameraDevice* QCameraDevice_new2(const QCameraDevice* other) {
	 return new QCameraDevice(*other);
}

void QCameraDevice_OperatorAssign(QCameraDevice* self, const QCameraDevice* other) {
	self->operator=(*other);
}

bool QCameraDevice_OperatorEqual(const QCameraDevice* self, const QCameraDevice* other) {
	return self->operator==(*other);
}

bool QCameraDevice_OperatorNotEqual(const QCameraDevice* self, const QCameraDevice* other) {
	return self->operator!=(*other);
}

bool QCameraDevice_IsNull(const QCameraDevice* self) {
	return self->isNull();
}

libqt_string QCameraDevice_Id(const QCameraDevice* self) {
	QByteArray _qb = self->id();
	libqt_string _str;
	_str.len = _qb.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _qb.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QCameraDevice_Description(const QCameraDevice* self) {
	QString _ret = self->description();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

bool QCameraDevice_IsDefault(const QCameraDevice* self) {
	return self->isDefault();
}

int QCameraDevice_Position(const QCameraDevice* self) {
	return self->position();
}

libqt_list QCameraDevice_PhotoResolutions(const QCameraDevice* self) {
	auto _ret = self->photoResolutions();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

libqt_list QCameraDevice_VideoFormats(const QCameraDevice* self) {
	auto _ret = self->videoFormats();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

int QCameraDevice_CorrectionAngle(const QCameraDevice* self) {
	return self->correctionAngle();
}

void QCameraDevice_Delete(QCameraDevice* self) {
    delete self;
}

File: binding/src/multimedia/libqcameradevice.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qcameraformat.html
pub const qcameraformat = struct {

    /// New constructs a new QCameraFormat object.
    pub fn New() QtC.QCameraFormat {
        return qtc.QCameraFormat_new();
    }


    /// New2 constructs a new QCameraFormat object.
    pub fn New2(other: ?*anyopaque) QtC.QCameraFormat {
        return qtc.QCameraFormat_new2(@ptrCast(other));
    }


    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QCameraFormat_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    pub fn PixelFormat(self: ?*anyopaque, ) i32 {
        return qtc.QCameraFormat_PixelFormat(@ptrCast(self));
    }

    pub fn Resolution(self: ?*anyopaque, ) QtC.QSize {
        return qtc.QCameraFormat_Resolution(@ptrCast(self));
    }

    pub fn MinFrameRate(self: ?*anyopaque, ) f32 {
        return qtc.QCameraFormat_MinFrameRate(@ptrCast(self));
    }

    pub fn MaxFrameRate(self: ?*anyopaque, ) f32 {
        return qtc.QCameraFormat_MaxFrameRate(@ptrCast(self));
    }

    pub fn IsNull(self: ?*anyopaque, ) bool {
        return qtc.QCameraFormat_IsNull(@ptrCast(self));
    }

    pub fn OperatorEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QCameraFormat_OperatorEqual(@ptrCast(self), @ptrCast(other));
    }

    pub fn OperatorNotEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QCameraFormat_OperatorNotEqual(@ptrCast(self), @ptrCast(other));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QCameraFormat_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qcameradevice.html
pub const qcameradevice = struct {

    /// New constructs a new QCameraDevice object.
    pub fn New() QtC.QCameraDevice {
        return qtc.QCameraDevice_new();
    }


    /// New2 constructs a new QCameraDevice object.
    pub fn New2(other: ?*anyopaque) QtC.QCameraDevice {
        return qtc.QCameraDevice_new2(@ptrCast(other));
    }


    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QCameraDevice_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    pub fn OperatorEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QCameraDevice_OperatorEqual(@ptrCast(self), @ptrCast(other));
    }

    pub fn OperatorNotEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QCameraDevice_OperatorNotEqual(@ptrCast(self), @ptrCast(other));
    }

    pub fn IsNull(self: ?*anyopaque, ) bool {
        return qtc.QCameraDevice_IsNull(@ptrCast(self));
    }

    pub fn Id(self: ?*anyopaque, , allocator: std.mem.Allocator) []u8 {
        const _bytearray: qtc.libqt_string = qtc.QCameraDevice_Id(@ptrCast(self));
defer qtc.libqt_string_free(&_bytearray);
const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("qcameradevice.Id: Memory allocation failed");
@memcpy(_ret, _bytearray.data[0.._bytearray.len]);
return  _ret;
    }

    pub fn Description(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QCameraDevice_Description(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qcameradevice.Description: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn IsDefault(self: ?*anyopaque, ) bool {
        return qtc.QCameraDevice_IsDefault(@ptrCast(self));
    }

    pub fn Position(self: ?*anyopaque, ) i32 {
        return qtc.QCameraDevice_Position(@ptrCast(self));
    }

    pub fn PhotoResolutions(self: ?*anyopaque, ) []const u8 {
        return qtc.QCameraDevice_PhotoResolutions(@ptrCast(self));
    }

    pub fn VideoFormats(self: ?*anyopaque, ) []const u8 {
        return qtc.QCameraDevice_VideoFormats(@ptrCast(self));
    }

    pub fn CorrectionAngle(self: ?*anyopaque, ) i32 {
        return qtc.QCameraDevice_CorrectionAngle(@ptrCast(self));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QCameraDevice_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/cameradevice.html#types
pub const enums = struct {
    pub const Position = enum {
        pub const UnspecifiedPosition: i32 = 0;
        pub const BackFace: i32 = 1;
        pub const FrontFace: i32 = 2;
    };

};
